blog/_posts/2025-09-05-pair-llmd-and-rag-on-aks.md

```diff
@@ -0,0 +1,179 @@
+* **KV Cache (Key-Value Cache)**: Stores key and value tensors from the prefill stage’s attention computations, enabling the decode stage to reuse these results for faster token generation, with reduced computational overhead.
+
+## Benefits of llm-d and its intersection with RAG
+
+The llm-d framework, built on open-source technologies like [vLLM](https://docs.vllm.ai/en/latest/index.html), [Gateway API Inference Extension](https://gateway-api-inference-extension.sigs.k8s.io/) (GAIE), and [NIXL](https://github.com/ai-dynamo/nixl), is a Kubernetes-native distributed inference serving stack for serving LLMs at scale. As detailed in the [llm-d documentation](https://llm-d.ai/docs/architecture), it provides several key benefits, particularly when paired with RAG workflows, which often involve long context to keep LLMs up to date.
+
+![llm-d architecture](/assets/images/pair-llmd-and-rag-on-aks/llmd-arch.png)
+
+|llm-d feature|What it does|Benefits to RAG workflows|
+|--|--|--|
+|Prefill/Decode (P/D) Disaggregation|Separates the compute-heavy prefill stage (KV cache building) from the decode stage (autoregressive token generation) on dedicated GPUs, where each GPU pool can be independently scheduled and scaled|Optimizes throughput for long contexts; prevents resource contention for concurrent requests when processing long RAG queries; enables flexible scaling and lowers time-to-first-token (TTFT) and improves overall token output time (TPOT)|
+|Intelligent routing (via Gateway API Inference Extension's Endpoint Picker)|Schedules requests based not only on server queue length and available KV cache size, but also prefix cache hit probability. Maximize the reuse of KV cache for queries with overlapping system prompts and common context retrieved from the vector store|Minimizes latency; enables efficient reuse of KV cache for similar or repeated contexts; handles RAG queries of mixed context length and structure robustly|
+|Disaggregated KV cache management|Offloads KV cache across local or remote stores, with orchestration from decode-sidecar and advanced cache eviction policies for efficient memory use|Supports much longer input contexts and multiple concurrent sessions with reduced GPU memory overhead, enabling scalability for large RAG pipelines|
+
+## Let’s get started: KAITO RAGEngine backed by llm-d with P/D Disaggregation
+
+With the [prerequisites](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#prerequisites) covered, this guide will dive into the creation of two distinct but related endpoints:
+
+1. **Inference Endpoint**: an OpenAI API compatible inference service in Kubernetes, created by the llm-d stack. Jump to the [llm-d inference endpoint](https://github.com/kaito-project/kaito-cookbook/tree/master/examples/ragengine-llm-d#inference-endpoint) section in our GitHub repository to set it up.
```

Comment:
Should also leave a note to ensure the reader has access to meta-llama/Llama-3.1-8B-Instruct by visiting https://huggingface.co/meta-llama/Llama-3.1-8B-Instruct to be whitelisted.